File: nlp/dev.py
import subprocess
import sys
import os
import signal
import threading
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler
import time
import signal

FORCE_EXIT_TIMEOUT = 5
RESTART_DEBOUNCE = 0.5
LOG_PREFIX = "[dev]"

class DevHandler(FileSystemEventHandler):
    def __init__(self, command: list):
        self.command = command
        self.restart_debounce_timer = None
        self.lock = threading.Lock()
        self.start_process()

    def stop_process(self):
        if self.process.poll() is None:
            try :
                self.process.send_signal(signal.SIGTERM)
                self.process.wait(FORCE_EXIT_TIMEOUT)
            except subprocess.TimeoutExpired:
                print(f"{LOG_PREFIX} Process did not stop in time. Force killing...")
                self.process.kill()

    def start_process(self):
        self.process = subprocess.Popen(self.command, start_new_session=True)

    def handle_event(self, event):
        if not event.src_path.endswith('.py'):
            return

        def restart_debounce_callback():
            print(f"{LOG_PREFIX} File change detected. restarting process...")
            self.stop_process()
            self.start_process()

        with self.lock:
            if self.restart_debounce_timer is not None:
                self.restart_debounce_timer.cancel()
            self.restart_debounce_timer = threading.Timer(
                RESTART_DEBOUNCE,
                restart_debounce_callback
            )
            self.restart_debounce_timer.start()

    def on_created(self, event):
        self.handle_event(event)

    def on_modified(self, event):
        self.handle_event(event)


def main():
    path = os.path.join(os.path.dirname(__file__), "src")
    command = [sys.executable, "-u", "-m", "src.main"] 
    stop_triggered = False

    event_handler = DevHandler(command)
    observer = Observer()
    observer.schedule(event_handler, str(path), recursive=True)
    observer.start()

    def stop():
        nonlocal stop_triggered
        stop_triggered = True
        observer.stop()
        event_handler.stop_process()

    def signal_handler(_sig, _frame):
        print(f"{LOG_PREFIX} Stopping development server...")
        stop()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    print(f"{LOG_PREFIX} Starting development server...")

    try:
        while not stop_triggered:
            time.sleep(1)
    except Exception as e:
        print(f"{LOG_PREFIX} An error occurred: {e}")
        stop()
    observer.join()

if __name__ == "__main__":
    main()

File: core-server/dev.js
const { spawn } = require("child_process");
const chokidar = require("chokidar");
const path = require("path");

const FORCE_EXIT_TIMEOUT = 5000;
const RESTART_DEBOUNCE = 500;
const LOG_PREFIX = "[dev]";
const ENTRY = path.resolve(__dirname, "start.js");

let watcher = null;
let subProcess = null;
let stopTriggered = false;
let restartTimeout = null;

function startProcess() {
  return new Promise((resolve) => {
    if (stopTriggered) {
      resolve();
      return;
    }

    const buildProcess = spawn("npm", ["run", "--silent", "build"], {
      stdio: "inherit",
    });

    buildProcess.on("exit", (code) => {
      if (stopTriggered) {
        resolve();
        return;
      }

      if (code !== 0) {
        resolve();
        return;
      }

      const startProcess = spawn("node", [ENTRY], {
        stdio: "inherit",
        detached: true,
      });

      startProcess.on("exit", () => {
        subProcess = null;
      });

      subProcess = startProcess;
      resolve();
    });
  });
}

function stopProcess() {
  return new Promise((resolve) => {
    const currentProcess = subProcess;
    subProcess = null;

    if (!currentProcess) {
      resolve();
      return;
    }

    const forceExitTimeout = setTimeout(() => {
      console.warn(
        `${LOG_PREFIX} Process did not stop in time. Force killing...`,
      );
      currentProcess.removeAllListeners("exit");
      currentProcess.kill("SIGKILL");
      resolve();
    }, FORCE_EXIT_TIMEOUT);

    currentProcess.removeAllListeners("exit");
    currentProcess.on("exit", () => {
      clearTimeout(forceExitTimeout);
      resolve();
    });
    currentProcess.kill("SIGTERM");
  });
}

let starting = false;
let stopping = false;

function handleFileChange() {
  if (stopTriggered || stopping || starting) {
    return;
  }

  if (restartTimeout) {
    clearTimeout(restartTimeout);
  }

  restartTimeout = setTimeout(async () => {
    if (stopTriggered) {
      return;
    }

    console.info(`${LOG_PREFIX} File change detected. restarting process...`);

    stopping = true;
    await stopProcess();
    stopping = false;

    if (stopTriggered) {
      return;
    }

    starting = true;
    await startProcess();
    starting = false;
  }, RESTART_DEBOUNCE);
}

async function stop() {
  stopTriggered = true;
  console.info(`${LOG_PREFIX} Stopping server...`);

  if (watcher) {
    try {
      await watcher.close();
    } catch (error) {
      console.error(`${LOG_PREFIX} Error closing watcher: ${error}`);
    }
  }

  await stopProcess();
}

process.on("SIGINT", stop);
process.on("SIGTERM", stop);

console.info(`${LOG_PREFIX} Starting server...`);
startProcess().then(() => {
  if (stopTriggered) {
    return;
  }

  watcher = chokidar.watch(["src", "../shared"], {
    persistent: true,
  });

  watcher
    .on("error", (error) =>
      console.error(`${LOG_PREFIX} Watcher error: ${error}`),
    )
    .on("ready", () => {
      watcher
        .on("add", handleFileChange)
        .on("change", handleFileChange)
        .on("unlink", handleFileChange);
    });
});
